### PR TITLE
Integrated the ippc_shmem and vx_producer vx_consumer changes

### PR DESCRIPTION
**Pull request by raphacan**

### CODE REVIEW

File: include/VX/vx.h [hunk 1]
```diff
@@ -43,6 +43,10 @@
  */
 #define VX_MAX_REFERENCE_NAME (64)
 
+#define VX_MAX_NUM_REFERENCES   (32U)
+
+#define VX_MAX_BUFFER_POOL_SIZE (5U)
```
Comment (raphacan) [inline include/VX/vx.h]: not used to be removed

File: include/VX/vx.h [hunk 2]
```diff
@@ -43,6 +43,10 @@
  */
 #define VX_MAX_REFERENCE_NAME (64)
 
+#define VX_MAX_NUM_REFERENCES   (32U)
```
Comment (raphacan) [inline include/VX/vx.h]: is there not another define already available in the framework?

File: include/VX/vx_khr_safe_casts.h [hunk 3]
```diff
@@ -71,6 +71,10 @@ __MAKE_SAFE_CASTS__(threshold, Threshold)
 #ifdef VX_TYPE_USER_DATA_OBJECT
 __MAKE_SAFE_CASTS__(user_data_object, UserDataObject)
 #endif
+#if defined(LINUX) || defined(QNX)
```
Comment (raphacan) [inline include/VX/vx_khr_safe_casts.h]: why this protection? why don't we create a #ifdef VX_PRODUCER_CONSUMER_EXT define as #ifdef VX_TYPE_USER_DATA_OBJECT?

File: source/include/tivx_objects.h [hunk 4]
```diff
@@ -113,7 +113,15 @@ typedef struct _tivx_object_t
     /**< User Data Objects */
     vx_bool                 isUserDataObjectUse[TIVX_USER_DATA_OBJECT_MAX_OBJECTS];
     /**< Flag indicating if User data object is in use or not */
```
Comment (raphacan) [inline source/include/tivx_objects.h:115]: is the comment wrong?

File: source/include/vx_gw_common.h [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+    // flag set when metadata can be read by consumer
+    uint8_t metadata_valid;
+
+    // flag to inform consumer whether previous frame has been dropped by producer
+    uint8_t last_frame_dropped;
+
+    // size of metadata, copied after this struct in the buffer ID message
+    size_t metadata_size;
+} SProducerContent;
```
Comment (raphacan) [inline source/include/vx_gw_common.h]: we need to adapt the struct's names to the ovx style

File: source/include/vx_gw_common.h [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+//consumer->producer message (1->1)
+typedef struct
+{
+    uint32_t buffer_id;
+    uint32_t last_buffer;
+    uint32_t consumer_id;
+} SConsumerContent;
```
Comment (raphacan) [inline source/include/vx_gw_common.h]: we need to adapt the struct's names to the ovx style, maybe like you did below

File: source/include/vx_gw_common.h [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    // flag to inform consumer whether previous frame has been dropped by producer
+    uint8_t last_frame_dropped;
+
+    // size of metadata, copied after this struct in the buffer ID message
+    uint16_t metadata_size;
+
+} vx_gw_buff_id_msg;
+
+  /* may number of consumer for socket or IPP*/
+  #define OVXGW_NUM_CLIENTS (4U)
```
Comment (raphacan) [inline source/include/vx_gw_common.h]: put all the defines at the beginning of the file

File: source/framework/vx_producer.c [hunk 8]
```diff
@@ -0,0 +1,1336 @@
...
+                &num_items,
+                &ipc_message_parent,
+                (tivx_utils_ref_ipc_msg_t*)&ipc_message_item[0]);
+
+            buffid_message->num_items = num_items;
+
+            // send object array items data, if present
+            for (j = 0; j < num_items; j++)
+            {
+                memcpy(&buffid_message->ipc_msg[i][j], (void*)&ipc_message_item[j], sizeof(tivx_utils_ref_ipc_msg_t));
```
Comment (raphacan) [inline source/framework/vx_producer.c]: don't memcopy if you do need it: you have the pointer to the memory buffid_message where you have to write into. So pass this point to the export function.

File: source/framework/vx_producer.c [hunk 9]
```diff
@@ -0,0 +1,1336 @@
...
+            VX_PRINT(
+                VX_ZONE_INFO,
+                "PRODUCER %s: export of buffer successfull: %d of total: %d\n",
+                producer->name,
+                i + 1,
+                producer->numBufferRefsExport);
+        }
+
+        // send reference data, for object array this is final metadata
+        memcpy(&buffid_message->ipc_msg[i][j], &ipc_message_parent, sizeof(tivx_utils_ref_ipc_msg_t));
```
Comment (raphacan) [inline source/framework/vx_producer.c]: same here, do not memcopy if not needed. This is a waste of runtime and CPU load

File: source/framework/vx_producer.c [hunk 10]
```diff
@@ -0,0 +1,1336 @@
...
+#include <vx_internal.h>
+
+static vx_status set_buffer_status(vx_reference current_ref, producer_buffer_status status, vx_producer producer)
+{
+    uint8_t     buffer_id;
+    vx_bool     found                = vx_false_e;
+    vx_bool     forbidden_transition = vx_false_e;
+    const char* state2string[]       = {"IN_GRAPH", "LOCKED", "FREE"};
+
+    pthread_mutex_lock(&producer->buffer_mutex);
```
Comment (raphacan) [inline source/framework/vx_producer.c:26]: now that we are in ovx, why don't we use tivx_vx_mutex_t?

File: source/framework/vx_producer.c [hunk 11]
```diff
@@ -0,0 +1,1336 @@
...
+        {
+            VX_PRINT(
+                VX_ZONE_INFO, "PRODUCER %s: number of objects to exchange reached, set last object\n", producer->name);
+            buffer_desc_msg->last_reference = 1;
+        }
+
+        // send reference data, for object array this is final metadata
+        buffer_desc_msg->item_index = 0; // used only for object array items
+        memcpy(&buffer_desc_msg->ipc_msg, &ipc_message_parent, sizeof(tivx_utils_ref_ipc_msg_t));
+#ifdef defined(QNX)
```
Comment (raphacan) [inline source/framework/vx_producer.c]: socket write for QNX?

File: source/framework/vx_producer.c [hunk 12]
```diff
@@ -0,0 +1,1336 @@
...
+        status = socket_write(client->socket_fd, (uint8_t*)buffer_desc_msg, NULL, 0);
+#elif defined(x86_64)
```
Comment (raphacan) [inline source/framework/vx_producer.c]: why not LINUX here? we have to choose.
Socket are for Linux, if this is linux arm or x86, it does not make a difference
QNX works with IPPC

File: source/framework/vx_producer.c [hunk 13]
```diff
@@ -0,0 +1,1336 @@
...
+    client_context* client   = (client_context*)clientPtr;
+    vx_producer producer = (vx_producer)data;
+
+    uint8_t           message_buffer[SOCKET_MAX_MSG_SIZE];
+    vx_gw_hello_msg* consumer_message;
+
+    int32_t client_num = -1;
+    int32_t status     = VXGW_STATUS_SUCCESS;
+
+    while (1)
```
Comment (raphacan) [inline source/framework/vx_producer.c:526]: change the 1 by true? Or another variable to handle the shutdown?

File: source/framework/vx_producer.c [hunk 14]
```diff
@@ -0,0 +1,1336 @@
...
+                    VX_PRINT(
+                        VX_ZONE_ERROR, "PRODUCER %s: Reference %u, could not be set to FREE \n", producer->name, i);
+                }
+            }
+        }
+    }
+
+    pthread_mutex_unlock(&producer->client_mutex);
+
+    return;
```
Comment (raphacan) [inline source/framework/vx_producer.c]: it's a void function, you do not need the return

File: source/framework/vx_producer.c [hunk 15]
```diff
@@ -0,0 +1,1336 @@
...
+                                //if one of the consumer is newly connected
+                                if (producer->consumers_list[i].state == PROD_STATE_CLI_CONNECTED)
+                                {
+                                    VX_PRINT(
+                                        VX_ZONE_INFO,
+                                        "PRODUCER %s: send buffer metadata for consumer %u \n", producer->name, i);
+
+                                    // send init message with backchannel port and buffer metadata
+                                    buffid_message->buffer_meta = 0x12345678;
+                                    SIppcPortMap * port_map = ippc_get_port_by_recv_index(&producer->m_shmem_ctx, i);//shm_context, i);
```
Comment (raphacan) [inline source/framework/vx_producer.c:882]: comment to be removed?

File: source/framework/vx_producer.c [hunk 16]
```diff
@@ -0,0 +1,1336 @@
...
+
+                                    buffid_message->consumer_id = i; // index is enough to set up connection on the other side
+                                    buffid_message->backchannel_port = port_map->m_port_id; // offset by 1
+
+                                    // set the meta....
```
Comment (raphacan) [inline source/framework/vx_producer.c]: remove the comment

File: include/VX/vx.h (hunk 1 above)
Comment (htd1kor) [reply include/VX/vx.h]: Removed

File: include/VX/vx.h (hunk 2 above)
Comment (htd1kor) [reply include/VX/vx.h]: This is temporary change

File: include/VX/vx_khr_safe_casts.h (hunk 3 above)
Comment (htd1kor) [reply include/VX/vx_khr_safe_casts.h]: Changed the macro to VX_TYPE_PRODUCER and VX_TYPE_CONSUMER

File: source/include/tivx_objects.h (hunk 4 above)
Comment (htd1kor) [reply source/include/tivx_objects.h:115]: No, this is expected

File: source/include/vx_gw_common.h (hunk 5 above)
Comment (htd1kor) [reply source/include/vx_gw_common.h]: Done

File: source/include/vx_gw_common.h (hunk 6 above)
Comment (htd1kor) [reply source/include/vx_gw_common.h]: Done

File: source/include/vx_gw_common.h (hunk 7 above)
Comment (htd1kor) [reply source/include/vx_gw_common.h]: Done

File: source/framework/vx_producer.c (hunk 8 above)
Comment (htd1kor) [reply source/framework/vx_producer.c]: Done

File: source/framework/vx_producer.c (hunk 9 above)
Comment (htd1kor) [reply source/framework/vx_producer.c]: Done

File: source/framework/vx_producer.c (hunk 11 above)
Comment (htd1kor) [reply source/framework/vx_producer.c]: Done

File: source/framework/vx_producer.c (hunk 12 above)
Comment (htd1kor) [reply source/framework/vx_producer.c]: Done

File: source/framework/vx_producer.c (hunk 14 above)
Comment (htd1kor) [reply source/framework/vx_producer.c]: Done

File: source/framework/vx_producer.c (hunk 15 above)
Comment (htd1kor) [reply source/framework/vx_producer.c:882]: Done

File: source/framework/vx_producer.c (hunk 16 above)
Comment (htd1kor) [reply source/framework/vx_producer.c]: Done

File: source/framework/vx_producer.c (hunk 10 above)
Comment (htd1kor) [reply source/framework/vx_producer.c:26]: Will be taken care in the followup PR

File: source/framework/vx_producer.c (hunk 13 above)
Comment (htd1kor) [reply source/framework/vx_producer.c:526]: Reviewed and agreed

**Pull request merged**